cdap-data-fabric/src/main/java/co/cask/cdap/spi/data/table/TableSpecification.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.spi.data.table;
+
+import co.cask.cdap.spi.data.InvalidFieldException;
+import co.cask.cdap.spi.data.StructuredTable;
+import co.cask.cdap.spi.data.table.field.FieldType;
+import co.cask.cdap.spi.data.table.field.Fields;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * Defines the specification of a {@link StructuredTable}.
+ * The table specification contains:
+ * <ul>
+ *   <li>table Id - specifies the name of the table</li>
+ *   <li>fields - the schema of the table, consists of the column names and their types</li>
+ *   <li>primaryKeys - the primary key for each row</li>
+ *   <li>indexes - the columns to index on. Only one column can be part of an index</li>
+ * </ul>
+ */
+public final class TableSpecification {
```

Comment:
Similar here, shall we call this `StructuredTableSpecification`?

Reply:
Done